packages/custom_lint/lib/src/v2/server_to_client_channel.dart

```diff
@@ -90,33 +104,50 @@ void _writePackageConfigForTempProject(
         continue;
       }
 
-      final currentPackage = packageMap[package.name];
+      final currentPackageWithContextRoot = packageMap[package.name];
+      final currentPackage = currentPackageWithContextRoot?.package;
 
       if (currentPackage != null && currentPackage.root != package.root) {
-        throw StateError(
-          '''
-Two ContextRoots depend on ${package.name} but use different version,
-therefore custom_lint does not know which one to pick.
-- ${package.root}
-- ${currentPackage.root}
-''',
+        packagesInMultipleContextRoots.add(
+          _PackageAndContextRootPair(
+            currentPackageWithContextRoot!,
+            _PackageAndContextRoot(package, contextRoot),
+          ),
         );
       }
 
-      packageMap[package.name] = package;
+      packageMap[package.name] = _PackageAndContextRoot(package, contextRoot);
     }
   }
+  if (packagesInMultipleContextRoots.isNotEmpty) {
+    throw StateError(
+      'The following packages are included in multiple ContextRoots:\n'
```

Comment:
It probably doesn't make sense to keep talking about ContextRoot in the error message.
It's technically correct, but it's an analyzer term and we'll confuse people.

We should have a more user-friendly message explaining what happened (cf https://github.com/invertase/dart_custom_lint/issues/89#issuecomment-1445217359).
And it should contain a solution to the problem.

Such as saying "run 'pub upgrade' in the following packages :<...>"